pkg/cli/server/server.go

```diff
@@ -364,8 +364,18 @@ func run(app *cli.Context, cfg *cmds.Server, leaderControllers server.CustomCont
 		if err != nil {
 			return errors.Wrap(err, "cannot configure default cluster-dns address")
 		}
+
 		serverConfig.ControlConfig.ClusterDNS = clusterDNS
 		serverConfig.ControlConfig.ClusterDNSs = []net.IP{serverConfig.ControlConfig.ClusterDNS}
+
+		if len(serverConfig.ControlConfig.ServiceIPRanges) > 1 {
+			// dualStack. We must pin the ipv6 address
+			extraDNS, err := utilsnet.GetIndexedIP(serverConfig.ControlConfig.ServiceIPRanges[1], 10)
+			if err != nil {
+				return errors.Wrap(err, "problems calculating extraDNS")
+			}
+			serverConfig.ControlConfig.ClusterDNSs = append(serverConfig.ControlConfig.ClusterDNSs, extraDNS)
+		}
 	} else {
```

Comment:
This should be covered by upgrade testing, but it is good to think about.

If the apply controller doesn't handle it properly, we may have to add some additional logic to delete and re-create the service so that we can change the addresses. I don't **think** this should be a problem, as I believe should be allowed to change clusterIPs when switching from single-stack to dual-stack? 

If someone does as described here, and intentionally tries to change the cluster-dns address after the fact, I think they are responsible for doing the proper cleanup to allow that to work.

Reply:
I think that updating important config parameters in a running cluster is something that we currently don't support in k3s. For example: `cluster-dns`,`service-cidr`, `cluster-cidr` or `flannel-backend`. If there is enough interest, we could try to design that feature and prototype it. In any case, I'd say this is out of the scope of this PR

Reply:
Ah, sorry, I think you two got me wrong.

Let's say I started a cluster with `k3s server --service-cidr 10.43.0.0/16,2001:cafe:42:1::/112 --cluster-dns 10.43.0.10`. Based on the current PR, the kube-dns service in the `coredns.yaml` will be rendered as:

```yaml
clusterIP: 10.43.0.10
clusterIPs: [10.43.0.10]
ipFamilyPolicy: PreferDualStack
```

Which will result a [`Service`](https://kubernetes.io/docs/reference/kubernetes-api/service-resources/service-v1/) resource similar to this:

```yaml
apiVersion: v1
kind: Service
metadata:
  name: kube-dns
  namespace: kube-system
spec:
  clusterIP: 10.43.0.10
  clusterIPs:
  - 10.43.0.10
  - 2001:cafe:42:1::62c1 # A random IPv6 address
  ipFamilyPolicy: PreferDualStack
```

So far so good.

Then we manually `kubectl apply -f coredns.yaml` again without any changes (this will happen every time k3s restarts, right?). I think we'll get this error:

```
The Service "kube-dns" is invalid: spec.ipFamilyPolicy: Invalid value: "PreferDualStack": must be 'SingleStack' to release the secondary cluster IP
```

This is because the `clusterIPs` field cannot be changed once set.

I know that we didn't make any changes to the `coredns.yaml` itself, but we were actually trying to release the secondary cluster IP (that random IPv6 address) of the kube-dns service without knowing it. Because the rendered field is always `clusterIPs: [10.43.0.10]`.

Reply:
Thanks @aofei for the explanation, indeed I was not understanding what you meant. You have a good point. To be safe, we probably should use `SingleStack` for kube-dns service if cluster-dns flag is set with a singlestack value. It should be easy to code. 

Moreover, if afterwards the user realizes a mistake was done and wanted dual-stack kube-dns, the user could change the value of cluster-dns adding the extra IP and restart k3s. Even if it means changing an important config on a running cluster, I think we could support this case because the main DNS IP would not change and we would just be adding a new extra IP